core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateUnnest.java

```diff
@@ -0,0 +1,402 @@
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.CorrelatedJoinNode;
+import io.trino.sql.planner.plan.EnforceSingleRowNode;
+import io.trino.sql.planner.plan.FilterNode;
+import io.trino.sql.planner.plan.JoinNode.Type;
+import io.trino.sql.planner.plan.LimitNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.PlanNodeId;
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.sql.planner.plan.RowNumberNode;
+import io.trino.sql.planner.plan.UnnestNode;
+import io.trino.sql.tree.Cast;
+import io.trino.sql.tree.ComparisonExpression;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.FunctionCall;
+import io.trino.sql.tree.GenericLiteral;
+import io.trino.sql.tree.IfExpression;
+import io.trino.sql.tree.IsNullPredicate;
+import io.trino.sql.tree.NullLiteral;
+import io.trino.sql.tree.QualifiedName;
+import io.trino.sql.tree.StringLiteral;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.matching.Pattern.nonEmpty;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static io.trino.sql.analyzer.TypeSignatureProvider.fromTypes;
+import static io.trino.sql.analyzer.TypeSignatureTranslator.toSqlType;
+import static io.trino.sql.planner.iterative.rule.Util.restrictOutputs;
+import static io.trino.sql.planner.optimizations.QueryCardinalityUtil.isScalar;
+import static io.trino.sql.planner.plan.JoinNode.Type.INNER;
+import static io.trino.sql.planner.plan.JoinNode.Type.LEFT;
+import static io.trino.sql.planner.plan.Patterns.CorrelatedJoin.correlation;
+import static io.trino.sql.planner.plan.Patterns.CorrelatedJoin.filter;
+import static io.trino.sql.planner.plan.Patterns.correlatedJoin;
+import static io.trino.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static io.trino.sql.tree.ComparisonExpression.Operator.GREATER_THAN;
+import static io.trino.sql.tree.ComparisonExpression.Operator.LESS_THAN_OR_EQUAL;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This rule decorrelates plans with correlated UnnestNode and optional EnforceSingleRowNode,
+ * optional LimitNode and optional projections in the subquery.
+ * <p>
+ * The rule finds correlated UnnestNode in CorrelatedJoinNode's subquery and folds
+ * them into UnnestNode representing INNER or LEFT JOIN UNNEST.
+ * It transforms plans, where:
+ * - CorrelatedJoinNode is INNER or LEFT on true
+ * - UnnestNode in subquery is based only on correlation symbols
+ * - UnnestNode in subquery is INNER or LEFT without filter
+ * <p>
+ * Transforms:
+ * <pre>
+ * - CorrelatedJoin (INNER or LEFT) on true, correlation(c)
+ *      - Input (a, c)
+ *     [- EnforceSingleRow]
+ *          [- Limit (5)]
```

Comment:
what about TopN instead Limit? It feels like more useful case.

Reply:
Ok - I see that at planning time this is same node :)

Reply:
A test would be nice to have though.

Reply:
Ignore email notification regarding this comment. I had a discussion with myself :)